### PR TITLE
Google auth

### PR DESCRIPTION
**Pull request by adarsh1001111**
## Description
Added Google Auth Using Firebase for SignUp and Login.

## Related Issue
#277 

## Screenshots :
Google SignUp Button added on the SignUp page:
Screenshot 2024-05-12 at 1 46 15 AM

Clicking the Google SignUp Button Opens the Google Auth Page having your Gmail accounts:
Screenshot 2024-05-12 at 1 46 58 AM

Selecting the account to SignUp with:
Screenshot 2024-05-12 at 1 47 18 AM

SignUp in progress:
Screenshot 2024-05-12 at 1 47 28 AM

User Signed Up :
Screenshot 2024-05-12 at 1 47 36 AM

User Details(name,email) captured via Google Account:
Screenshot 2024-05-12 at 1 47 46 AM

User Successfully LoggedOut via LogOut Button on the UserDetails Page:
Screenshot 2024-05-12 at 1 47 54 AM

## Checklist:
- [x] I have performed a self-review of my code
- [x] I have added/updated relevant documentation (if needed)
- [x] I have tested the changes locally and they function as expected
- [x] I have ensured my code follows the project's coding standards

## Additional Notes
For Now, the SignUp through Google automatically sets the Password to -"0000" and the Phone number to -"0000000000"(10 digits) Since these info can't be provided only via the Google Auth, there should be an option to change the Phone Number and Password further after the login in another issue for this. Also as many websites provide a password via email if logged in through Google auth can be done.

### CODE REVIEW

File: public/images/search.png
Comment (adarsh1001111) [inline public/images/search.png:1]: The google logo added to make the button look good for Google singup/login

File: src/pages/user/UserRegistration.jsx [hunk 1]
```diff
@@ -16,150 +21,261 @@ function LoginRegisterForm() {
     isLogin = !isLogin;
   }
 
+  const googleButtonStyle = {
```
Comment (adarsh1001111) [inline src/pages/user/UserRegistration.jsx:24]: added css for google button

File: src/pages/user/UserRegistration.jsx [hunk 2]
```diff
@@ -16,150 +21,261 @@ function LoginRegisterForm() {
...
+                      )
+                    }
+                  />
+                  {error && (
+                    <p className="absolute top-[-25px] w-screen tracking-wide text-red-500 font-semibold text-center">
+                      {error}
+                    </p>
+                  )}
+                </div>
+                <ReactiveButton
```
Comment (adarsh1001111) [inline src/pages/user/UserRegistration.jsx:157]: Google SignUp Button added below the Form

File: src/pages/user/UserRegistration.jsx [hunk 3]
```diff
@@ -16,150 +21,261 @@ function LoginRegisterForm() {
...
+                    }
+                  />
+                  {error && (
+                    <p className="absolute w-screen top-[-40px] tracking-wide text-red-500 font-semibold text-center">
+                      {error}
+                    </p>
+                  )}
+                </div>
+                <div className="w-screen relative mt-10 h-fit flex justify-center">
+                  <ReactiveButton
```
Comment (adarsh1001111) [inline src/pages/user/UserRegistration.jsx:232]: Google SignIn Button at the Login Page

File: src/utils/Functions/userAuthService.js [hunk 4]
```diff
@@ -2,67 +2,114 @@ import axios from "axios";
...
  * @param {Function} setError - A function to set the error message in the component state.
  * @param {Function} setButtonState - A function to set the state of the login button (idle/loading/success/error).
  * @returns {void}
  */
-export const login = async (email, password, setError, setButtonState) => {
-    // Validate the email address
-    if (!isValidEmail(email)) {
-        setError("Enter a valid email address.");
-        return;
+export const googleLogin = async (setError, setButtonState) => {
```
Comment (adarsh1001111) [inline src/utils/Functions/userAuthService.js:17]: Google Login Auth Function

File: src/utils/Functions/userAuthService.js [hunk 5]
```diff
@@ -72,50 +119,96 @@ export const login = async (email, password, setError, setButtonState) => {
...
+      setError(error.response.data.error);
+    } else if (error.message === "Network Error") {
+      setError("Network error.Please check your internet connection.");
+    } else {
+      setError("An error occurred while registering.");
     }
+  }
+};
+
+export const googleSignup = async (setError, setButtonState) => {
```
Comment (adarsh1001111) [inline src/utils/Functions/userAuthService.js:174]: Google SignUp Auth function added

**Comment by rishicds** [comment]
Hi @adarsh1001111 , good job but can you please make the button rounded so it fits in with the rest of website? Thank you.

File: public/firebase-messaging-sw.js
Comment (codewarnab) [inline public/firebase-messaging-sw.js:1]: do not push your firebase config files to github i have update the .gitignore file syncg your branch and keep the config chanages to firebase locally

File: public/index.html
Comment (codewarnab) [inline public/index.html:1]: whats the benefit of adding '/' at the end ?

File: src/pages/user/UserRegistration.jsx
Comment (codewarnab) [inline src/pages/user/UserRegistration.jsx:1]: keep the chanages of this files locally too

File: src/firebase.js
Comment (codewarnab) [inline src/firebase.js:1]: keep the chanages of this file locally

File: src/pages/user/UserRegistration.jsx [hunk 6]
```diff
@@ -4,10 +4,15 @@ import InputField from "../../Components/shared/InputField";
 import ReactiveButton from "reactive-button";
 import Background from "../../Components/shared/Background";
 import {
+  googleLogin,
+  googleSignup,
   login,
   registration,
 } from "../../utils/Functions/userAuthService";
 import LoginTextLink from "../../Components/shared/LoginTextLink";
+import { AlignHorizontalLeft } from "@mui/icons-material";
```
Comment (codewarnab) [inline src/pages/user/UserRegistration.jsx:13]: where have you used this two imports ?

File: src/pages/user/UserRegistration.jsx [hunk 7]
```diff
@@ -16,150 +21,261 @@ function LoginRegisterForm() {
...
   const [name, setname] = useState("");
   const [phone_number, setPhone_number] = useState("");
   const [email, setEmail] = useState("");
   const [password, setPassword] = useState("");
   const [error, setError] = useState("");
   const [state, setButtonState] = useState("idle");
- useEffect(() => {
-  setError("");
- },[isLogin])
+  const [googleButtonState, setGoogleButtonState] = useState("idle");
```
Comment (codewarnab) [inline src/pages/user/UserRegistration.jsx:57]: where have you used this state ?

File: src/pages/user/UserRegistration.jsx [hunk 8]
```diff
@@ -16,150 +21,261 @@ function LoginRegisterForm() {
     isLogin = !isLogin;
   }
 
+  const googleButtonStyle = {
+    borderRadius: "40px",
+    display: "flex",
+    borderRadius: "4px",
```
Comment (codewarnab) [inline src/pages/user/UserRegistration.jsx:27]: fix this duplicate styles borderRadius

**Comment by codewarnab** [comment]
image
i am facing this error ( ihave allowed popup still please test it properply google login ) 
also you cant send a arbitary number  keep it blank if the backend allows else need to modify backend to fit google login properly

File: public/firebase-messaging-sw.js
Comment (adarsh1001111) [reply public/firebase-messaging-sw.js:1]: no, actually when i forked the repo the firebase file already contained these config so i tested the web app and then restored the config to the previous before making the PR.

File: public/index.html
Comment (adarsh1001111) [reply public/index.html:1]: it was just a part of my CSS extension , when i save any code , it just formats the code text in concise manner,we can change that

Comment (codewarnab) [reply public/index.html:1]: ok than change it

**Comment by adarsh1001111** [comment]
> image i am facing this error ( ihave allowed popup still please test it properply google login ) also you cant send a arbitary number keep it blank if the backend allows else need to modify backend to fit google login properly

it says auth config not found means what i did with firebase.js (i exported the auth ) and imported in the register page so that it could work

**Comment by adarsh1001111** [comment]
you can just add another file for firebase.js and write the firebase config of your paws project of authentication and then use getAuth and export that auth, @codewarnab , also backend didnt allowed "" null string in phone and pass thats why i did "0 casing"

**Comment by adarsh1001111** [comment]
> Hi @adarsh1001111 , good job but can you please make the button rounded so it fits in with the rest of website? Thank you.

i did it actually it was overridden by border radius 4px , i corrected it , its 40px rounded now

**Comment by adarsh1001111** [comment]
i have uploaded a video of my testing on my drive , view this once 
link: https://drive.google.com/file/d/1zbPozyHE8HgcKXumrMldP1SkeXUbYJYK/view?usp=drive_link

**Comment by adarsh1001111** [comment]
@codewarnab any updates ? if this PR is fine , i will switch to the backend for phone number and password for google auth sign in, please check if nothing is broken , till my knowledge everything works just the phone no. and pass has to be fixed from backend repo

**Comment by rishicds** [comment]
Hi @adarsh1001111 ,its okay, good job!

**Pull request merged**